imbi/slugify.py

```diff
@@ -0,0 +1,36 @@
+from __future__ import annotations
+
+import re
+import typing
+
+import pydantic
+
+Slug = typing.Annotated[str,
+                        pydantic.StringConstraints(
+                            pattern=r'^[a-z0-9][-_a-z0-9]*[a-z0-9]?$',
+                            min_length=1,
+                        )]
+
+
+def path_element(value: str) -> str:
+    """Generate a slug for an arbitrary string"""
+    value = re.sub(r'[\s_]+', '-', value)
+    value = re.sub(r'[^-a-z0-9]', '', value.lower())
+    return Slug(value.strip('-_'))
+
+
+def decode_path_slug(slug_or_id: str) -> tuple[int | None, str | None]:
```

Comment:
I think the output typing would actually be

```python
tuple[int, None] | tuple[None, str]
```

Reply:
doh ... missed this one